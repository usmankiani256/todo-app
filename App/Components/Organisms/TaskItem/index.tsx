import * as React from 'react'
import { View } from 'react-native'
import { TouchableRipple } from 'react-native-paper'
import useStyles from './styles'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
import { Colors } from 'Theme'
import { Text } from 'Components/Atoms'

export type TaskItemProps = {
  priority: 'high' | 'medium' | 'low'
  status: 'checked' | 'unchecked'
  description: string
  onToggle: () => void
  onPress: () => void
}

const TaskItem = (props: TaskItemProps) => {
  const { priority, status, description, onToggle, onPress } = props

  let indicatorColor =
    priority === 'high'
      ? Colors.priority.high
      : priority === 'medium'
      ? Colors.priority.medium
      : Colors.priority.low

  let taskStatus = status === 'checked' ? true : false

  const { container, indicator, text } = useStyles()

  return (
    <TouchableRipple
      testID="task-item-button"
      onPress={onPress}
      style={container}>
      <>
        <View
          style={{
            ...indicator,
            backgroundColor: indicatorColor,
            borderColor: indicatorColor,
          }}
        />
        <Icon
          testID="task-item-icon"
          onPress={onToggle}
          name={taskStatus ? 'checkbox-marked' : 'checkbox-blank-outline'}
          color={taskStatus ? Colors.primary : Colors.tintGrey}
          size={25}
          style={{ opacity: taskStatus ? 0.7 : 1 }}
        />
        <Text
          justify
          style={{
            ...text,
            textDecorationLine: taskStatus ? 'line-through' : undefined,
            opacity: taskStatus ? 0.7 : 1,
          }}>
          {description}
        </Text>
      </>
    </TouchableRipple>
  )
}

export default TaskItem
