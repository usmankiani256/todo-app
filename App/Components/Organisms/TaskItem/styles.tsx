import { Dimensions, Platform, StyleSheet } from 'react-native'
import { Colors } from 'Theme'

const { width } = Dimensions.get('window')

const useStyles = () =>
  StyleSheet.create({
    container: {
      minHeight: 50,
      width: width - 36,
      alignSelf: 'center',
      borderRadius: 10,
      borderWidth: 1,
      borderLeftWidth: 0,
      borderColor: Colors.tintGrey,
      flexDirection: 'row',
      alignItems: 'center',
      marginBottom: 15,
    },
    indicator: {
      height: '100%',
      borderWidth: 1,
      borderColor: Colors.tintGrey,
      width: 8,
      borderTopLeftRadius: 10,
      borderBottomLeftRadius: 10,
      marginRight: 10,
    },
    text: {
      paddingTop: 2,
      marginLeft: 10,
      marginRight: 50,
      marginVertical: 8,
    },
  })

export default useStyles
