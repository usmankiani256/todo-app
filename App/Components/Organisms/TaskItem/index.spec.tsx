import * as React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import TaskItem from './index'

describe('TaskItem Component', () => {
  it('renders high priority, checked and description', () => {
    const rendered: any = render(
      <TaskItem
        priority="high"
        status="checked"
        onToggle={() => {}}
        onPress={() => {}}
        description="Create header page component"
      />,
    ).toJSON()

    expect(rendered).toMatchSnapshot()
  })

  it('renders medium priority, unchecked', () => {
    const rendered: any = render(
      <TaskItem
        priority="medium"
        status="unchecked"
        onToggle={() => {}}
        onPress={() => {}}
        description="Create header page component"
      />,
    ).toJSON()

    expect(rendered).toMatchSnapshot()
  })

  it('renders low priority', () => {
    const rendered: any = render(
      <TaskItem
        priority="low"
        status="unchecked"
        onToggle={() => {}}
        onPress={() => {}}
        description="Create header page component"
      />,
    ).toJSON()

    expect(rendered).toMatchSnapshot()
  })

  it('onPress fires', () => {
    const mockFn = jest.fn()

    const { getByTestId } = render(
      <TaskItem
        priority="low"
        status="unchecked"
        onToggle={() => {}}
        onPress={mockFn}
        description="Create header page component"
      />,
    )

    fireEvent.press(getByTestId('task-item-button'))

    expect(mockFn).toBeCalledTimes(1)
  })

  it('onToggle fires', () => {
    const mockFn = jest.fn()

    const { getByTestId } = render(
      <TaskItem
        priority="low"
        status="unchecked"
        onToggle={mockFn}
        onPress={() => {}}
        description="Create header page component"
      />,
    )

    fireEvent.press(getByTestId('task-item-icon'))

    expect(mockFn).toBeCalledTimes(1)
  })
})
