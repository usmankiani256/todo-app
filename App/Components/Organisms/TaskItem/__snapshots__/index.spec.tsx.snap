// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem Component renders high priority, checked and description 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "borderColor": "#777b",
        "borderLeftWidth": 0,
        "borderRadius": 10,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginBottom": 15,
        "minHeight": 50,
        "width": 714,
      },
    ]
  }
  testID="task-item-button"
>
  <View
    style={
      Object {
        "backgroundColor": "#F87171",
        "borderBottomLeftRadius": 10,
        "borderColor": "#F87171",
        "borderTopLeftRadius": 10,
        "borderWidth": 1,
        "height": "100%",
        "marginRight": 10,
        "width": 8,
      }
    }
  />
  <Text
    allowFontScaling={false}
    onPress={[Function]}
    style={
      Array [
        Object {
          "color": "#31394F",
          "fontSize": 25,
        },
        Object {
          "opacity": 0.7,
        },
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="task-item-icon"
  >
    󰄲
  </Text>
  <Text
    style={
      Object {
        "color": "#3f3f3f",
        "fontFamily": "SofiaPro",
        "fontSize": 20,
        "lineHeight": 20,
        "marginLeft": 10,
        "marginRight": 50,
        "marginVertical": 8,
        "opacity": 0.7,
        "paddingTop": 2,
        "textAlign": "justify",
        "textDecorationLine": "line-through",
      }
    }
  >
    Create header page component
  </Text>
</View>
`;

exports[`TaskItem Component renders low priority 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "borderColor": "#777b",
        "borderLeftWidth": 0,
        "borderRadius": 10,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginBottom": 15,
        "minHeight": 50,
        "width": 714,
      },
    ]
  }
  testID="task-item-button"
>
  <View
    style={
      Object {
        "backgroundColor": "#3460DC",
        "borderBottomLeftRadius": 10,
        "borderColor": "#3460DC",
        "borderTopLeftRadius": 10,
        "borderWidth": 1,
        "height": "100%",
        "marginRight": 10,
        "width": 8,
      }
    }
  />
  <Text
    allowFontScaling={false}
    onPress={[Function]}
    style={
      Array [
        Object {
          "color": "#777b",
          "fontSize": 25,
        },
        Object {
          "opacity": 1,
        },
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="task-item-icon"
  >
    󰄱
  </Text>
  <Text
    style={
      Object {
        "color": "#3f3f3f",
        "fontFamily": "SofiaPro",
        "fontSize": 20,
        "lineHeight": 20,
        "marginLeft": 10,
        "marginRight": 50,
        "marginVertical": 8,
        "opacity": 1,
        "paddingTop": 2,
        "textAlign": "justify",
        "textDecorationLine": undefined,
      }
    }
  >
    Create header page component
  </Text>
</View>
`;

exports[`TaskItem Component renders medium priority, unchecked 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "borderColor": "#777b",
        "borderLeftWidth": 0,
        "borderRadius": 10,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginBottom": 15,
        "minHeight": 50,
        "width": 714,
      },
    ]
  }
  testID="task-item-button"
>
  <View
    style={
      Object {
        "backgroundColor": "#FBBF24",
        "borderBottomLeftRadius": 10,
        "borderColor": "#FBBF24",
        "borderTopLeftRadius": 10,
        "borderWidth": 1,
        "height": "100%",
        "marginRight": 10,
        "width": 8,
      }
    }
  />
  <Text
    allowFontScaling={false}
    onPress={[Function]}
    style={
      Array [
        Object {
          "color": "#777b",
          "fontSize": 25,
        },
        Object {
          "opacity": 1,
        },
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
    testID="task-item-icon"
  >
    󰄱
  </Text>
  <Text
    style={
      Object {
        "color": "#3f3f3f",
        "fontFamily": "SofiaPro",
        "fontSize": 20,
        "lineHeight": 20,
        "marginLeft": 10,
        "marginRight": 50,
        "marginVertical": 8,
        "opacity": 1,
        "paddingTop": 2,
        "textAlign": "justify",
        "textDecorationLine": undefined,
      }
    }
  >
    Create header page component
  </Text>
</View>
`;
